File: code/language/src/Checker/Unify.ml
(* IMPORTANT: It is expected that all free type variables are bound in the
 * prefix! If this is not true we will panic.
 *
 * The unification algorithm takes a prefix and two monotypes asserting that the
 * types are equivalent under the prefix. However, what makes unification extra
 * special is that at the same time if we encounter flexible bounds in the
 * prefix then we update our prefix so that the types become equivalent under
 * the prefix.
 *
 * In the terms of the [MLF thesis][1] unification is defined as:
 *
 * > Definition 4.1.1 (Unification): A prefix `Q'` unifies `t1` and `t2` under
 * > `Q` if and only if `Q ⊑ Q'` and `(Q') t1 ≡ t2` hold.
 *
 * Our implementation is a bit different then the algorithm proposed in the
 * thesis. Our algorithm _never fails_. Instead our algorithm returns a result.
 * If the result is ok then the two types are equivalent. That is `(Q') t1 ≡ t2`
 * holds. If our algorithm result is an error then the two types are _not_
 * equivalent. That is `(Q') t1 ≡ t2` does _not_ hold. The prefix after
 * unification will always be an instance of the prefix before unification. That
 * is `Q ⊑ Q'` always holds.
 *
 * The caller of the unification algorithm which depends on equivalent types for
 * soundness must handle the error locally.
 *
 * [1]: https://pastel.archives-ouvertes.fr/file/index/docid/47191/filename/tel-00007132.pdf *)
let rec unify prefix type1 type2 =
  match type1.Type.monotype_description, type2.Type.monotype_description with
  (* Variables with the same name unify without any further analysis. We rename
   * variables in type1 and type2 so that the same variable name does not
   * represent different bounds. *)
  | Variable { name = name1 }, Variable { name = name2 } when String.equal name1 name2 ->
    Ok ()

  (* Unify type1 with some other monotype. If the monotype is a variable then
   * we will perform some special handling. *)
  | Variable { name }, description2 -> (
    let bound1 = Prefix.lookup prefix name in
    match bound1.bound_type.polytype_description, description2 with
    (* If the bound for our variable is a monotype then recursively call `unify`
     * with that monotype. As per the normal-form monotype bound
     * rewrite rule. *)
    | Monotype type1, _ -> unify prefix type1 type2

    (* Two variables with different names were unified with one another. This
     * case is different because we need to merge the two variables together. *)
    | _, Variable { name = name2 } -> (
      let bound2 = Prefix.lookup prefix name2 in
      match bound2.bound_type.polytype_description with
      (* If the bound for our variable is a monotype then recursively call
       * `unify` with that monotype. As per the normal-form monotype bound
       * rewrite rule. Make sure we don’t fall into the next case where we try
       * to update the types to each other! *)
      | Monotype type2 -> unify prefix type1 type2

      (* Actually merge the two type variables together. *)
      | _ -> (
        let name1 = name in
        match unify_polytype prefix bound1.bound_type bound2.bound_type with
        | Error error -> Error error
        | Ok t ->
          let bound_kind = if (
            bound1.bound_kind = Flexible &&
            bound2.bound_kind = Flexible
          ) then Type.Flexible else Type.Rigid in
          Prefix.update2 prefix name1 name2 (Type.bound bound_kind t)
      )
    )

    (* Unify the polymorphic bound type with our other monotype. If the
     * unification is successful then update the type variable in our prefix to
     * our monotype. *)
    | _, _ -> (
      let type2 = Type.to_polytype type2 in
      match unify_polytype prefix bound1.bound_type type2 with
      | Error error -> Error error
      | Ok _ -> Prefix.update prefix name (Type.bound Rigid type2)
    )
  )

  (* Unify type2 with some other monotype. The other monotype is guaranteed
   * to not be a variable since that case would be caught by the match
   * case above. *)
  | _, Variable { name } -> (
    let bound2 = Prefix.lookup prefix name in
    match bound2.bound_type.polytype_description with
    (* If the bound for our variable is a monotype then recursively call `unify`
     * with that monotype. As per the normal-form monotype bound rewrite rule. *)
    | Monotype type2 -> unify prefix type1 type2

    (* Unify the polymorphic bound type with our other monotype. If the
     * unification is successful then update the type variable in our prefix to
     * our monotype. *)
    | _ -> (
      let type1 = Type.to_polytype type1 in
      match unify_polytype prefix type1 bound2.bound_type with
      | Error error -> Error error
      | Ok _ -> Prefix.update prefix name (Type.bound Rigid type1)
    )
  )

  (* Constant intrinsic types unify with each other no problem. *)
  | Boolean, Boolean -> Ok ()
  | Number, Number -> Ok ()

  (* Functions unify if the parameters and bodies unify.
   *
   * If both the unification of the parameters and bodies fail then we will
   * report two error diagnostics. However, we will only return the first error
   * from unify. *)
  | Function { parameter = parameter1; body = body1 },
    Function { parameter = parameter2; body = body2 } -> (
    let result1 = unify prefix parameter1 parameter2 in
    let result2 = unify prefix body1 body2 in
    match result1, result2 with
    | Error error, _ -> Error error
    | Ok (), Error error -> Error error
    | Ok (), Ok () -> Ok ()
  )

  (* Exhaustive match for failure case. Don’t use `_` since if we add a new type
   * we want an error telling us to add a case for that type to unification. *)
  | Boolean, _
  | Number, _
  | Function _, _ ->
    let type1 = Type.to_polytype type1 in
    let type2 = Type.to_polytype type2 in
    Error (Diagnostics.report_error (IncompatibleTypes { type1; type2 }))

(* Unifies two polytypes. When the two types are equivalent we return an ok
 * result with a type. This type is an instance of both our input types. That is
 * if `t1` and `t2` are our inputs and `t1 ≡ t2` holds then we return `t3` where
 * both `t1 ⊑ t3` and `t2 ⊑ t3` hold. *)
and unify_polytype prefix type1 type2 =
  match type1.Type.polytype_description, type2.Type.polytype_description with
  (* If either is bottom then return the other one. *)
  | Bottom, _ -> Ok type2
  | _, Bottom -> Ok type1

  (* If we have two monotypes then unify them. Don’t bother with creating a new
   * level or generalizing. *)
  | Monotype type1, Monotype type2 -> (
    match unify prefix type1 type2 with
    | Ok () -> Ok (Type.to_polytype type1)
    | Error error -> Error error
  )

  (* When two quantified types unify we instantiate their local bounds in the
   * prefix. We consider a monotype to be a quantified type with an empty bounds
   * list. We then unify the body of the two quantified types in the new prefix.
   * If unification is successful then we know that the two types we unified
   * must be equivalent. We then generalize type1 and return it. Since we know
   * our types to be equivalent it does not matter which one we return. If
   * unification returned an error then we also return that error. We do all
   * this in an isolated level so that we don’t quantify type variables that are
   * needed at an earlier level. *)

  | Quantify { bounds = bounds1; body = body1 }, Monotype type2 ->
    Prefix.level prefix (fun () -> (
      let type1 = Prefix.instantiate prefix bounds1 body1 in
      match unify prefix type1 type2 with
      | Ok () -> Ok (Prefix.generalize prefix type1)
      | Error error -> Error error
    ))

  | Monotype type1, Quantify { bounds = bounds2; body = body2 } ->
    Prefix.level prefix (fun () -> (
      let type2 = Prefix.instantiate prefix bounds2 body2 in
      match unify prefix type1 type2 with
      | Ok () -> Ok (Prefix.generalize prefix type1)
      | Error error -> Error error
    ))

  | Quantify { bounds = bounds1; body = body1 }, Quantify { bounds = bounds2; body = body2 } ->
    Prefix.level prefix (fun () -> (
      let type1 = Prefix.instantiate prefix bounds1 body1 in
      let type2 = Prefix.instantiate prefix bounds2 body2 in
      match unify prefix type1 type2 with
      | Ok () -> Ok (Prefix.generalize prefix type1)
      | Error error -> Error error
    ))

File: old/old/old/old/experimental/mlf-type-system/src/mlf/unify.ts
import {BindingMap} from '../utils/bindings';
import {Err, Ok, Result} from '../utils/result';

import {Diagnostics, Reported} from './diagnostics';
import {generalize} from './generalize';
import {Prefix, PrefixError} from './prefix';
import {Bound, Monotype, Polytype, Type} from './type';

export type UnifyError =
  | PrefixError
  | {
      readonly kind: 'IncompatibleTypes';
      readonly actual: Monotype;
      readonly expected: Monotype;
    };

/**
 * Implements the unification algorithm from the [MLF thesis][1].
 *
 * In theory, the paper defines unification as:
 *
 * > Definition 4.1.1 (Unification): A prefix `Q'` unifies `t1` and `t2` under
 * > `Q` if and only if `Q ⊑ Q'` and `(Q') t1 ≡ t2` hold.
 *
 * So every mutation to the prefix must maintain the invariant `Q ⊑ Q'`. In our
 * implementation an unbounded `a` type variable represent the theory for the
 * prefix entry `a ≥ ⊥`. According to the instance relation every polymorphic
 * type is an instance of bottom (`⊥`).
 *
 * As an optimization we also mutate type variables in prefix to an equivalent
 * type when unification proves it is safe.
 *
 * So `Q ⊑ Q'` holds, but what about `(Q') t1 ≡ t2`? This one is interesting
 * since for an improved developer experience we don’t want to stop type
 * checking if any component of `t1` is not equivalent to the same component
 * in `t2`. For example, `unify((a), number → string, boolean → a)`. Here
 * `number ≢ boolean`. The result in the [MLF thesis][1] would be to fail type
 * checking. In our implementation, however, this unification will emit an error
 * when it sees `number ≢ boolean` but will also continue and update `a` to
 * `string` resulting in the prefix `Q'` as `(a = string)`. Here `Q ⊑ Q'` holds,
 * but `(Q') t1 ≡ t2` does not!
 *
 * How do we maintain soundness? Our implementation of unify will return an
 * error. It is then expected that the caller of unify handle that error! So if
 * the caller expects `(Q') t1 ≡ t2` to hold and it does not then the caller
 * should insert a runtime error in the program where the assertion failed.
 *
 * Consider the program:
 *
 * ```
 * let x = 42 in
 * let y = (x: string) in
 * ...
 * ```
 *
 * Here `number ≢ string`, so we convert this code to:
 *
 * ```
 * let x = 42 in
 * let y = error "number ≢ string" in
 * ...
 * ```
 *
 * At runtime this will halt execution of our program. It is yet to be shown in
 * theory that this strategy is type sound.
 *
 * [1]: https://pastel.archives-ouvertes.fr/file/index/docid/47191/filename/tel-00007132.pdf
 */
export function unify<Diagnostic>(
  diagnostics: Diagnostics<UnifyError | Diagnostic>,
  prefix: Prefix,
  actual: Monotype,
  expected: Monotype
): Reported<UnifyError | Diagnostic> | undefined {
  return unifyMonotype(diagnostics, prefix, actual, expected);
}

/**
 * Unifies two monotypes.
 */
function unifyMonotype<Diagnostic>(
  diagnostics: Diagnostics<UnifyError | Diagnostic>,
  prefix: Prefix,
  actual: Monotype,
  expected: Monotype
): Reported<UnifyError | Diagnostic> | undefined {
  // If the actual and expected type variable names match then we know the two
  // types are identical and we don’t need to recursively unify. This only
  // works because we rename variables during instantiation so names
  // don’t conflict!
  if (
    actual.description.kind === 'Variable' &&
    expected.description.kind === 'Variable' &&
    actual.description.name === expected.description.name
  ) {
    return undefined;
  }

  // If the bound of actual is a monomorphic then let us recursively call
  // unify with that type.
  let actualVariable:
    | {readonly name: string; readonly bound: Bound}
    | undefined;
  if (actual.description.kind === 'Variable') {
    const entry = prefix.lookup(actual.description.name);
    // If no type variable exists with this name then report an error.
    if (entry.kind === 'Err') {
      return diagnostics.report(entry.value);
    }
    // Return the bound from this block and check for shortcuts.
    actualVariable = {name: actual.description.name, bound: entry.value.bound};
    if (actualVariable.bound.type.description.kind === 'Variable') {
      const actualType = actualVariable.bound.type as Monotype;
      return unifyMonotype(diagnostics, prefix, actualType, expected);
    }
  }

  // If the bound of expected is a monomorphic then let us recursively call
  // unify with that type.
  let expectedVariable:
    | {readonly name: string; readonly bound: Bound}
    | undefined;
  if (expected.description.kind === 'Variable') {
    const entry = prefix.lookup(expected.description.name);
    // If no type variable exists with this name then report an error.
    if (entry.kind === 'Err') {
      return diagnostics.report(entry.value);
    }
    // Return the bound from this block and check for shortcuts.
    expectedVariable = {
      name: expected.description.name,
      bound: entry.value.bound,
    };
    if (expectedVariable.bound.type.description.kind === 'Variable') {
      const expectedType = expectedVariable.bound.type as Monotype;
      return unifyMonotype(diagnostics, prefix, actual, expectedType);
    }
  }

  // If both actual and expected are variables then unify their bounds. Also
  // unify the two types together at the very end.
  if (actualVariable !== undefined && expectedVariable !== undefined) {
    const result = unifyPolytype(
      diagnostics,
      prefix,
      actualVariable.bound.type,
      expectedVariable.bound.type
    );
    // If there was an error then don’t update our prefix. Only return the error
    // and carry on.
    if (result.kind === 'Err') {
      return result.value;
    } else {
      // Update actual to the new type and update expected to point at actual.
      // If we get an error in the process then return it.
      const type = result.value;
      const error1 = prefix.update(actualVariable.name, Type.rigidBound(type));
      if (error1 !== undefined) return diagnostics.report(error1);
      const error2 = prefix.update(
        expectedVariable.name,
        Type.rigidBound(Type.variable(actualVariable.name))
      );
      if (error2 !== undefined) return diagnostics.report(error2);
      return undefined;
    }
  }

  // If only actual was a variable then unify its type bound. If unification
  // succeeds then update the actual variable in the prefix.
  if (actualVariable !== undefined) {
    const actualType = actualVariable.bound.type;
    const result = unifyPolytype(diagnostics, prefix, actualType, expected);
    if (result.kind === 'Err') {
      return result.value;
    } else {
      const type = result.value;
      const error = prefix.update(actualVariable.name, Type.rigidBound(type));
      if (error !== undefined) return diagnostics.report(error);
      return undefined;
    }
  }

  // If only expected was a variable then unify its type bound. If unification
  // succeeds then update the expected variable in the prefix.
  if (expectedVariable !== undefined) {
    const expectedType = expectedVariable.bound.type;
    const result = unifyPolytype(diagnostics, prefix, actual, expectedType);
    if (result.kind === 'Err') {
      return result.value;
    } else {
      const type = result.value;
      const error = prefix.update(expectedVariable.name, Type.rigidBound(type));
      if (error !== undefined) return diagnostics.report(error);
      return undefined;
    }
  }

  // Matching constants unify.
  if (
    (actual.description.kind === 'Boolean' &&
      expected.description.kind === 'Boolean') ||
    (actual.description.kind === 'Number' &&
      expected.description.kind === 'Number') ||
    (actual.description.kind === 'String' &&
      expected.description.kind === 'String')
  ) {
    return undefined;
  }

  // Functions unify their parameter and body types.
  if (
    actual.description.kind === 'Function' &&
    expected.description.kind === 'Function'
  ) {
    // NOTE: We switch `expected` and `actual` when we unify the function
    // parameter since the function parameter is contravariant.
    const diagnostic1 = unifyMonotype(
      diagnostics,
      prefix,
      expected.description.parameter,
      actual.description.parameter
    );
    const diagnostic2 = unifyMonotype(
      diagnostics,
      prefix,
      actual.description.body,
      expected.description.body
    );
    return diagnostic1 || diagnostic2;
  }

  // If two types are not compatible then report a diagnostic and return.
  return diagnostics.report({
    kind: 'IncompatibleTypes',
    actual,
    expected,
  });
}

/**
 * Unifies two polytypes. Returns a type which both should be made equivalent to
 * if not already.
 */
function unifyPolytype<Diagnostic>(
  diagnostics: Diagnostics<UnifyError | Diagnostic>,
  prefix: Prefix,
  actual: Polytype,
  expected: Polytype
): Result<Polytype, Reported<UnifyError | Diagnostic>> {
  // If either is bottom then return the other.
  if (actual.description.kind === 'Bottom') return Ok(expected);
  if (expected.description.kind === 'Bottom') return Ok(actual);

  // Increment the level. We are about to create some type variables which will
  // need to be generalized.
  prefix.incrementLevel();

  // Merge the prefixes of actual and expected into our global prefix. We also
  // rename type variables when necessary to avoid collisions.
  const actualMonotype = mergePrefix(prefix, actual);
  const expectedMonotype = mergePrefix(prefix, expected);

  // If our type is of the form `∀a.⊥` then we won’t have handled the bottom
  // type above. Instead the bottom type will have been hiding out under
  // quantifications. We don’t have a sensible way to handle bottom types in
  // this case, so panic. We assume that the types being unified are in normal
  // form anyway so panicking is fine.
  if (actualMonotype === undefined || expectedMonotype === undefined) {
    throw new Error('Unsupported quantified bottom type.');
  }

  // Unify the underlying monotypes.
  const error = unifyMonotype(
    diagnostics,
    prefix,
    actualMonotype,
    expectedMonotype
  );

  if (error === undefined) {
    // Generalize the actual type in the current level. Because unification did
    // not produce an error we know that actual and expected are equivalent. So
    // we arbitrarily pick actual but it doesn’t matter which one we pick.
    const type = generalize(prefix, actualMonotype);
    // Decrement the level. We are done creating type variables.
    prefix.decrementLevel();
    // Return the type we generalized.
    return Ok(type);
  } else {
    // Decrement the level. We are done creating type variables. We don’t end up
    // using any of the type variables at this level though because we erred!
    prefix.decrementLevel();
    return Err(error);
  }
}

/**
 * Merges the prefix of a polytype into the global prefix. Returns the unwrapped
 * monotype or nothing if the provided polytype was the bottom type.
 */
function mergePrefix(prefix: Prefix, type: Polytype): Monotype | undefined {
  const substitutions = new BindingMap<string, Monotype | undefined>();

  // For all quantification bounds...
  while (type.description.kind === 'Quantify') {
    const {name, bound} = type.description;
    let newBound = bound;
    // If we have some substitutions apply them in our bound’s type.
    if (!substitutions.isEmpty()) {
      const newType = substitutePolytype(bound.type);
      newBound = {kind: newBound.kind, type: newType};
    }
    // Add our bound to the prefix. It is possible that a type variable with
    // this name already exists in the prefix.
    const newName = prefix.add(name, newBound);
    // If the type variable added to the prefix has a different name then let us
    // add a substitution for our type variable.
    //
    // Consider `∀a.∀(a = a).∀(a = a). ...` If we later see the same name then
    // our prefix will generate a new name and we will override the previous
    // substitution in our bindings map.
    //
    // Also note since we shallowly look through quantification bounds we don’t
    // need to pop substitutions.
    if (newName !== name) {
      substitutions.push(name, Type.variable(newName));
    }
    type = type.description.body;
  }

  // If we have the bottom type then return undefined.
  if (type.description.kind === 'Bottom') return undefined;

  // Substitute the monotype and return it.
  return substituteMonotype(type as Monotype);

  /**
   * Apply our substitutions to the provided polytype. Locally shadowed
   * variables will not be substituted. Will return the type if substitutions
   * are not needed.
   */
  function substitutePolytype(type: Polytype): Polytype {
    if (!needsSubstitution(type)) return type;

    switch (type.description.kind) {
      case 'Quantify': {
        const name = type.description.name;
        const bound = {
          kind: type.description.bound.kind,
          type: substitutePolytype(type.description.bound.type),
        };
        substitutions.push(name, undefined);
        const body = substitutePolytype(type.description.body);
        substitutions.pop();
        return Type.quantify(name, bound, body);
      }

      case 'Bottom':
        return type;

      default:
        return substituteMonotype(type as Monotype);
    }
  }

  /**
   * Apply our substitutions to the provided monotype. Will return the type if
   * substitutions are not needed.
   */
  function substituteMonotype(type: Monotype): Monotype {
    if (!needsSubstitution(type)) return type;

    switch (type.description.kind) {
      case 'Variable': {
        const newType = substitutions.get(type.description.name);
        return newType || type;
      }

      case 'Function': {
        return Type.function_(
          substituteMonotype(type.description.parameter),
          substituteMonotype(type.description.body)
        );
      }

      case 'Boolean':
      case 'Number':
      case 'String':
        return type;

      default:
        const never: never = type.description;
        return never;
    }
  }

  /**
   * Given the provided substitutions map does our type need a substitution?
   */
  function needsSubstitution(type: Polytype): boolean {
    const freeVariables = Type.getFreeVariables(type);
    // Iterate over the smaller set.
    if (freeVariables.size <= substitutions.distinctKeysCount()) {
      for (const variable of freeVariables) {
        const newType = substitutions.get(variable);
        if (newType !== undefined) return true;
      }
    } else {
      for (const [variable, newType] of substitutions.distinctEntries()) {
        if (freeVariables.has(variable) && newType !== undefined) {
          return true;
        }
      }
    }
    return false;
  }
}
